[quant] Add quantized::leaky_relu that takes scale/zero_point as input

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45885 [quant][fx][graphmode] Add support for sigmoid
* #45884 [quant][eagermode] Add support for sigmoid module
* #45883 [quant] Add quantized Sigmoid module
* #45882 [quant] Add quantized::sigmoid that take output_scale/output_zero_point as input
* #45714 [quant][graphmode][fx] Add warning for unsupported case
* #45712 [quant][graphmode][fx][eagermode] Add leaky relu support in quantization workflows
* #45711 [quant] Add quantized LeakyReLU module
* **#45702 [quant] Add quantized::leaky_relu that takes scale/zero_point as input**

Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24067681](https://our.internmc.facebook.com/intern/diff/D24067681)

jerryzh168 committed:
[quant] Add quantized::leaky_relu that takes scale/zero_point as input

Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantized::leaky_relu that takes scale/zero_point as input"


Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24067681](https://our.internmc.facebook.com/intern/diff/D24067681)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantized::leaky_relu that takes scale/zero_point as input"


Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24067681](https://our.internmc.facebook.com/intern/diff/D24067681)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantized::leaky_relu that takes scale/zero_point as input"


Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24067681](https://our.internmc.facebook.com/intern/diff/D24067681)

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qrelu.cpp:
```diff
@@ -170,8 +170,24 @@ class QRelu6 final {
+class QLeakyRelu final {
+ public:
+  static Tensor run(Tensor self, Scalar negative_slope, bool inplace, double output_scale, int64_t output_zero_point) {
+    // inplace argument is ignored now, TODO:support inplace
```
Should there be a warning or an error if inplace=True?

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qrelu.cpp:
```diff
@@ -113,7 +113,7 @@ Tensor& leaky_relu_out_quantized_cpu(Tensor& result, const Tensor& self,
 }
 
-Tensor heaky_relu_quantized_cpu(const Tensor& self, Scalar negval) {
+Tensor leaky_relu_quantized_cpu(const Tensor& self, Scalar negval) {
```
For my understanding: When is this called compared to QLeakyRelu::run?

jerryzh168 replied:
this is called by the original leaky_relu op that works on both float and quantized tensors and it does not need quantization parameter, QLeakyRelu::run is called by quantized::leaky_relu, which only works on quantized tensor and needs output quantization parameters

jerryzh168 committed:
Update on "[quant] Add quantized::leaky_relu that takes scale/zero_point as input"


Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24067681](https://our.internmc.facebook.com/intern/diff/D24067681)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantized::leaky_relu that takes scale/zero_point as input"


Summary:
https://github.com/pytorch/pytorch/issues/45593

Previously quantized leaky_relu does not require observation and just inherits
the quantization parameters from input, but that does not work very well in qat
This PR added a quantized::leaky_relu that has observation for output and it will
become the default leaky_relu that our quantization tools produce (eager/graph mode)

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24067681](https://our.internmc.facebook.com/intern/diff/D24067681)

[ghstack-poisoned]
